internal/dependency/manager.go

```diff
@@ -152,147 +89,102 @@ func (dm *DependendcyManager) Validate(ctx context.Context, piManifest *v1alpha1
 	}, nil
 }
 
-func (dm *DependendcyManager) CheckConflict(existingVersionStr string, dependency v1alpha1.Dependency) (*Conflict, error) {
-	existingVersion, err := semver.NewVersion(existingVersionStr)
-	if err != nil {
-		return nil, err
-	}
-	requiredRange, err := semver.NewConstraint(dependency.Version)
+// NewGraph constructs a DependencyGraph from all packages returned by clientAdapter.ListPackages
+func (dm *DependendcyManager) NewGraph(ctx context.Context) (*graph.DependencyGraph, error) {
+	pkgs, err := dm.clientAdapter.ListPackages(ctx)
 	if err != nil {
 		return nil, err
 	}
-	if ok := requiredRange.Check(existingVersion); !ok {
-		return &Conflict{
-			Actual: PackageWithVersion{
-				Name:    dependency.Name,
-				Version: existingVersionStr,
-			},
-			Required: PackageWithVersion{
-				Name:    dependency.Name,
-				Version: dependency.Version,
-			},
-		}, nil
-	} else {
-		return nil, nil
-	}
-}
-
-func (dm *DependendcyManager) createRequirement(dependency v1alpha1.Dependency) (*PackageWithVersion, error) {
-	requirement := &PackageWithVersion{
-		Name: dependency.Name,
-	}
-	if dependency.Version == "" {
-		if latest, err := dm.repoAdapter.GetLatestVersion("", dependency.Name); err != nil {
-			return nil, err
+	g := graph.NewGraph()
+	for _, pkg := range pkgs.Items {
+		var deps []v1alpha1.Dependency
+		installedVersion := pkg.Spec.PackageInfo.Version
+		if !pkg.DeletionTimestamp.IsZero() {
+			// A package that is currently being deleted is added to the graph, but in a state representing
+			// "not installed"
+			installedVersion = ""
 		} else {
-			requirement.Version = latest
+			if pi, err := dm.clientAdapter.GetPackageInfo(ctx, pkg.Status.OwnedPackageInfos[0].Name); err != nil {
+				return nil, err
+			} else {
+				if pi.Status.Manifest != nil {
+					deps = pi.Status.Manifest.Dependencies
+				}
+			}
 		}
-	} else {
-		if maxCompatible, err := dm.repoAdapter.GetMaxVersionCompatibleWith("", dependency.Name, dependency.Version); err != nil {
+		if err := g.Add(pkg.Name, installedVersion, deps, len(pkg.OwnerReferences) == 0); err != nil {
 			return nil, err
-		} else {
-			requirement.Version = maxCompatible
 		}
 	}
-	return requirement, nil
-}
-
-type dependentTuple struct {
-	Pkg        *v1alpha1.Package
-	Dependency v1alpha1.Dependency
+	return g, nil
 }
 
-func (dm *DependendcyManager) IsUpdateAllowed(ctx context.Context, pkg *v1alpha1.Package, version string) (Conflicts, error) {
-	dependents, err := dm.GetDependents(ctx, pkg, GetDependentsOptions{IncludeDependencies: true})
-	if err != nil {
+// addWithDependencies adds the specified manifest to the graph and calls itself for every dependency.
+// This way, transitive dependencies are supported as well.
+func (dm *DependendcyManager) addWithDependencies(
+	g *graph.DependencyGraph,
+	manifest v1alpha1.PackageManifest,
+	version string,
+	transitive bool,
```

Comment:
I think calling this flag `transitive` is slightly confusing, but as a caller I might expect this to mean: "please add this manifest and all its transitive dependencies". Since its only an internal function it might not hurt that much. The only alternatives I could think of now are `isTransitive`or `isDependency`

Reply:
Are the semantics clearer now that this part of the feature actually works as intended? 

I'm not sure about the benefit of `isTransitive` vs `transitive` tbh :smile: